### ISSUE TITLE
[2.0 provider] Unable to run EF Core specifications tests in .NET 4.6.1 class library

### ISSUE THREAD

**Issue by ErikEJ** [dotnet/efcore#9046]
Trying to run functional tests with preview2 from a .NET 4.6.1 class library (SQLCE functional tests)
```
Exception message:
     System.IO.FileLoadException : Could not load file or assembly 'System.ValueTuple, Version=0.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
      Stack Trace:
           at Microsoft.EntityFrameworkCore.Metadata.Internal.TableMapping.GetTableMappings(IModel model)
```
This is blocking me from updating the SQLCE EF Core provider to EF Core 2.0

### Steps to reproduce
Run the functional tests:
 ..\..\packages\xunit.runner.console.2.2.0\tools\xunit.console.exe bin\Debug\EntityFramework.SqlServerCompact.Legacy.FunctionalTests.dll

### Further technical details
EF Core version: 2.0 preview2 (from Nuget.org feed)
Database Provider: SQL CE
Operating system: Win 10 1703
IDE: VS 2017 15.3 preview 3

Branch: https://github.com/ErikEJ/EntityFramework.SqlServerCompact/tree/prev2-prep

**Comment by bricelam** [comment]
You may want to switch from using `packages.config` to [PackageReference](https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files). In my experience things generally work better with .NET Standard that way.

**Comment by ErikEJ** [comment]
How do I do that with an existing project?

**Comment by ErikEJ** [comment]
Found Nate's guide: http://www.natemcmaster.com/blog/2017/03/09/vs2015-to-vs2017-upgrade/

**Comment by ErikEJ** [comment]
Tried converting to new csproj format - Got this error now - I only intend to work on .NET 4.6.1 - does that mean I MUST install the 2.0 SDK?

Error		The version of Microsoft.NET.Sdk used by this project is insufficient to support .NET Standard 2.0 which is required by this project's references.  Please install version 2.0 or higher of the .NET Core SDK.	EntityFramework.SqlServerCompact35.FunctionalTests	C:\Program Files (x86)\Microsoft Visual Studio\Preview\Community\MSBuild\Microsoft\Microsoft.NET.Build.Extensions\Microsoft.NET.Build.Extensions.NETFramework.targets  47

This is top of my csproj:
´´´´
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <AssemblyName>EntityFramework.SqlServerCompact.Legacy.FunctionalTests</AssemblyName>
    <TargetFramework>net461</TargetFramework>
  </PropertyGroup>
´´´´

**Comment by bricelam** [comment]
That's more than necessary. You can use "old" csproj just with `PackageReference` items instead of `Reference` items.

For example, instead of this...
```XML
<ItemGroup>
    <Reference Include="Microsoft.EntityFrameworkCore.Relational.Specification.Tests, Version=2.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60, processorArchitecture=MSIL">
      <HintPath>..\..\packages\Microsoft.EntityFrameworkCore.Relational.Specification.Tests.2.0.0-preview2-25099\lib\net46\Microsoft.EntityFrameworkCore.Relational.Specification.Tests.dll</HintPath>
    </Reference>
    <!-- (and it's dozens of dependencies) -->
</ItemGroup>
```
...you can delete the `packages.config` file and change to this:
```XML
<ItemGroup>
    <PackageReference Include="Microsoft.EntityFrameworkCore.Relational.Specification.Tests">
      <Version>2.0.0-preview2-25099</Version>
    </Reference>
```

**Comment by ErikEJ** [comment]
Great, let me try that!

**Comment by ErikEJ** [comment]
Getting this: 

1>C:\Program Files (x86)\Microsoft Visual Studio\Preview\Community\MSBuild\Microsoft\NuGet\15.0\Microsoft.NuGet.targets(178,5): error : Your project.json doesn't have a runtimes section. You should add '"runtimes": { "win": { } }' to your project.json and then re-run NuGet restore.

Should I install the 2.0 SDK?

**Comment by bricelam** [comment]
Hmm... You shouldn't need it, but does adding the following help?

```XML
<PropertyGroup>
  <RuntimeIdentifier>win7-x86</RuntimeIdentifier>
</PropertyGroup>
```

**Comment by ErikEJ** [comment]
Nope!

**Comment by bricelam** [comment]
Did you close and re-open the solution? NuGet (in VS) may be in a bad state.

**Comment by ErikEJ** [comment]
Yes, no dice...

**Comment by bricelam** [comment]
Well, scratch that idea then. Keep the `packages.config` files.

Try deleting the `App.config` files in the tests and running `Add-BindingRedirects` from PMC.

Also ensure you're targeting .NET Framework 4.6.1 everywhere.

**Comment by ErikEJ** [comment]
Tried that - no luck at all, back to square one.

Add-BindingRedirect did not add any app.config at all.

**Comment by ErikEJ** [comment]
Sharing fusion log:

*** Assembly Binder Log Entry  (03-07-2017 @ 21:29:11) ***

The operation failed.
Bind result: hr = 0x80131040. No description available.

Assembly manager loaded from:  C:\Windows\Microsoft.NET\Framework64\v4.0.30319\clr.dll
Running under executable  C:\Code\EntityFramework.SqlServerCompact\packages\xunit.runner.console.2.2.0\tools\xunit.console.exe
--- A detailed error log follows. 

=== Pre-bind state information ===
LOG: DisplayName = System.ValueTuple, Version=0.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
 (Fully-specified)
LOG: Appbase = file:///C:/Code/EntityFramework.SqlServerCompact/test/EntityFramework.SqlServerCompact35.FunctionalTests/bin/Debug
LOG: Initial PrivatePath = NULL
LOG: Dynamic Base = NULL
LOG: Cache Base = C:\Users\Erik\AppData\Local\Temp\19655a91-eed4-4c0d-aa1b-34c6a70b7f13
LOG: AppName = 19655a91-eed4-4c0d-aa1b-34c6a70b7f13
Calling assembly : netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51.
===
LOG: This bind starts in default load context.
LOG: Using application configuration file: C:\Code\EntityFramework.SqlServerCompact\packages\xunit.runner.console.2.2.0\tools\xunit.console.exe.Config
LOG: Using host configuration file: 
LOG: Using machine configuration file from C:\Windows\Microsoft.NET\Framework64\v4.0.30319\config\machine.config.
LOG: Post-policy reference: System.ValueTuple, Version=0.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
LOG: GAC Lookup was unsuccessful.
LOG: Attempting download of new URL file:///C:/Code/EntityFramework.SqlServerCompact/test/EntityFramework.SqlServerCompact35.FunctionalTests/bin/Debug/System.ValueTuple.DLL.
LOG: Assembly download was successful. Attempting setup of file: C:\Code\EntityFramework.SqlServerCompact\test\EntityFramework.SqlServerCompact35.FunctionalTests\bin\Debug\System.ValueTuple.dll
LOG: Entering download cache setup phase.
LOG: Assembly Name is: System.ValueTuple, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
WRN: Comparing the assembly name resulted in the mismatch: Major Version
ERR: The assembly reference did not match the assembly definition found.
ERR: Setup failed with hr = 0x80131040.
ERR: Failed to complete setup of assembly (hr = 0x80131040). Probing terminated.

*** Assembly Binder Log Entry  (03-07-2017 @ 21:29:11) ***

The operation failed.
Bind result: hr = 0x80131040. No description available.

Assembly manager loaded from:  C:\Windows\Microsoft.NET\Framework64\v4.0.30319\clr.dll
Running under executable  C:\Code\EntityFramework.SqlServerCompact\packages\xunit.runner.console.2.2.0\tools\xunit.console.exe
--- A detailed error log follows. 

=== Pre-bind state information ===
LOG: DisplayName = System.ValueTuple, Version=0.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
 (Fully-specified)
LOG: Appbase = file:///C:/Code/EntityFramework.SqlServerCompact/test/EntityFramework.SqlServerCompact35.FunctionalTests/bin/Debug
LOG: Initial PrivatePath = NULL
LOG: Dynamic Base = NULL
LOG: Cache Base = C:\Users\Erik\AppData\Local\Temp\19655a91-eed4-4c0d-aa1b-34c6a70b7f13
LOG: AppName = 19655a91-eed4-4c0d-aa1b-34c6a70b7f13
Calling assembly : netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51.
===
LOG: This bind starts in default load context.
LOG: Using application configuration file: C:\Code\EntityFramework.SqlServerCompact\packages\xunit.runner.console.2.2.0\tools\xunit.console.exe.Config
LOG: Using host configuration file: 
LOG: Using machine configuration file from C:\Windows\Microsoft.NET\Framework64\v4.0.30319\config\machine.config.
LOG: Post-policy reference: System.ValueTuple, Version=0.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
LOG: GAC Lookup was unsuccessful.
LOG: Attempting download of new URL file:///C:/Code/EntityFramework.SqlServerCompact/test/EntityFramework.SqlServerCompact35.FunctionalTests/bin/Debug/System.ValueTuple.DLL.
LOG: Assembly download was successful. Attempting setup of file: C:\Code\EntityFramework.SqlServerCompact\test\EntityFramework.SqlServerCompact35.FunctionalTests\bin\Debug\System.ValueTuple.dll
LOG: Entering download cache setup phase.
LOG: Assembly Name is: System.ValueTuple, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
WRN: Comparing the assembly name resulted in the mismatch: Major Version
ERR: The assembly reference did not match the assembly definition found.
ERR: Setup failed with hr = 0x80131040.
ERR: Failed to complete setup of assembly (hr = 0x80131040). Probing terminated.

**Comment by BradBarnich** [comment]
I was able to get it working with a binding redirect.

<dependentAssembly>
        <assemblyIdentity name="System.ValueTuple" publicKeyToken="cc7b13ffcd2ddd51" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.2.0" newVersion="4.0.2.0" />
</dependentAssembly>

**Comment by ErikEJ** [comment]
Thanks very much, Brad!  I will try it out - but not really a sustainable, long term solution, given that NuGet does not even add this binding redirect.

**Comment by ErikEJ** [comment]
Yes, tested the binding redirect, works.

**Comment by ajcvickers** [comment]
@ErikEJ Are you unblocked now?

If there are issues doing things that should work but don't seem to, then it would be great if you could file issues against NuGet.

**Comment by ErikEJ** [comment]
@ajcvickers Yes, I am able to run the tests with the binding redirect added manually, and the xunit console runner. As metioned here: https://github.com/aspnet/EntityFramework/issues/9053#issuecomment-312905678 I do not have any hopes of ever being able to ship a 2.0-preview2-final provider, sadly. Hope the issue gets fixed, and that there will be proper support for .NET 4.6.1 based providers by RTM....

**Comment by ajcvickers** [comment]
@ErikEJ Very sorry to hear that Erik, but it's not completely clear to me what it is that is blocking you. Is it that the development experience for external contributors is just too hard/inefficient, or there still something specific blocking you from releasing?

**Comment by BradBarnich** [comment]
Yeah, full framework support got a bit more tricky with preview2. I was able to get going with 

1. upping to .net 4.6.1 (was 4.6)
1. that binding redirect
1. referencing `NETStandard.Library.NETFramework`
1. `<ImplicitlyExpandDesignTimeFacades>false</ImplicitlyExpandDesignTimeFacades>`

**Comment by ErikEJ** [comment]
@ajcvickers 
1: Requiring Visual Studio 2017 15.3 preview 3 - this is acceptables/understandable an due to .NET Standard 2.0)
2: The System.ValueTuple binding redirect, which you must add manually - kind of showstopper (yes, I know I can do Nuget app.config tricks, but....)

**Comment by ajcvickers** [comment]
@ErikEJ So your're saying you can't release it because then you're customers would need to manually add the binding direct, and that's not acceptable?

**Comment by ErikEJ** [comment]
Yes!

**Comment by ajcvickers** [comment]
Okay, thanks. Your feedback is greatly appreciated and will be communicated up.

**Comment by ErikEJ** [comment]
Thanks for that, I fully understand that we are all in a state of transition to .NET Standard, and things are in flux

**Comment by ErikEJ** [comment]
@BradBarnich I is my understadning that 3 and 4 on your list are fixed by using VS 2017 15.3 preview 3, and yes, .NET 4.6.1 is now the minimum requirement for EF Core

**Comment by ajcvickers** [comment]
Closing this as no-repro. Please re-open if it still repros with latest VS and msbuild-based NuGet (PackageReference).

**Comment by weshaggard** [comment]
>@BradBarnich I is my understanding that 3 and 4 on your list are fixed by using VS 2017 15.3 preview 3, and yes, .NET 4.6.1 is now the minimum requirement for EF Core

Yes 3 and 4 shouldn't be needed with the new VS tools.

As for the binding redirects those will still be needed but you should be able to get them auto-generated for your test projects by ensuring the following properties are set in your project:

```
    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
    <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
```

Please try that out and see if it fixes your issue.

**Comment by ErikEJ** [comment]
@weshaggard Thanks for the info, all interesting. but cannot see how this improves the experience for devs creating a classic .exe project ?? Is the AutGenerate.... setting always true??

**Comment by divega** [comment]
Just so that @weshaggard doesn't have to post it twice, here is an excerpt from his answer at 
https://github.com/ErikEJ/EntityFramework.SqlServerCompact/issues/463#issuecomment-314212330: 
> Yes for any new .NET Framework .exe projects AutoGenerateBindingRedirects is set to true so unless they explicitly set it to false or are pulling forward an old project that didn't have it set then they will get the redirects automatically generated.

**Comment by thomaswolff** [comment]
Just wanted to leave a comment here in case someone developing an Azure Function App is having problems with setting up a binding redirect for System.ValueTuple. Read this blog post for a solution: https://codopia.wordpress.com/2017/07/21/how-to-fix-the-assembly-binding-redirect-problem-in-azure-functions/

**Comment by jculverwell** [comment]
@thomaswolff 

With regards to the article https://codopia.wordpress.com/2017/07/21/how-to-fix-the-assembly-binding-redirect-problem-in-azure-functions/  
Can you share what you config setting you were using to get System.ValueTuple working.   I'm using 

```
"BindingRedirects": "[ { \"ShortName\": \"System.ValueTuple\", \"RedirectToVersion\": \"4.4.0\", \"PublicKeyToken\": \"cc7b13ffcd2ddd51\" } ]"
```
and it is saying it can't find the assembly.

**Comment by jculverwell** [comment]
@thomaswolff 

Ignore my last message....this worked for me

```
"BindingRedirects": "[ { \"ShortName\": \"System.ValueTuple\", \"RedirectToVersion\": \"4.0.2.0\", \"PublicKeyToken\": \"cc7b13ffcd2ddd51\" } ]"
```

**Comment by philip-reed** [comment]
Hi,

I have been trying to follow each of the discussions about the System.ValueTuple issue but im stuggling to understand.  I get the same error @ErikEJ mentions at the start of this when i try to run a unit test that uses UseInMemoryDatabase for EF Core.  Each of my projects target Framework 4.7.

I have tried adding the following to my unit test project, but it doesnt appear to have any effect:
` <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
    <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>`

I have also tried putting the following in app.config of my test project:

`<dependentAssembly>
            <assemblyIdentity name="System.ValueTuple" publicKeyToken="cc7b13ffcd2ddd51" />
            <bindingRedirect oldVersion="0.0.0.0-4.6.0.0" newVersion="4.6" />
        </dependentAssembly>`

Have I put these in the wrong place or have i missed something else?

**Comment by ErikEJ** [comment]
@philip-reed That looks like the wrong version number!  Make sure to use VS 2017 15.3 or later

**Comment by philip-reed** [comment]
@ErikEJ 
Hi Erik, I am using VS 2017 15.3.2.  
Am I correct in thinking the binding redirect should be in the app.config of the unit test project?

I had tried various version numbers for System.ValueTuple when trying this, and 4.6 was just the latest attempt.  If i look in the bin folder of my web project (the one under test) i have the following for System.ValueTuple, and this is where i got the version number from:

![capture](https://user-images.githubusercontent.com/7798175/29716882-11f47224-89a5-11e7-927a-7c4ea10325c6.PNG)

**Comment by ErikEJ** [comment]
Product versuion != Assembly version
This is what I have in my 4.6.1 test project (why not stay at 4.6.1?)

´´´´
      <dependentAssembly>
        <assemblyIdentity name="System.ValueTuple" publicKeyToken="cc7b13ffcd2ddd51" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.2.0" newVersion="4.0.2.0" />
      </dependentAssembly>
´´´´

**Comment by philip-reed** [comment]
@ErikEJ I swear i tried that, but it is working now, thank you.

I started at version 4.6.1, and tried 4.7 to see if it resolved this issue, seems happy now either way.

Thanks again.

**Comment by VividDigital** [comment]
For anyone having issues with this, [here ](https://codopia.wordpress.com/2017/07/21/how-to-fix-the-assembly-binding-redirect-problem-in-azure-functions/)is a great workaround for Azure Functions

**Comment by codearoo** [comment]
We have this issue new, with .Net 2.0 Core. Using VS 2017 15.3.3.
Installed ValueTuple even with Nuget for the project wanting, but it's still bombing with the 
`System.ValueTuple, Version=0.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference.`

**Comment by codearoo** [comment]
Interesting... I searched for the cc7b13ffcd2ddd51 key, and it's finding it in app.config files as a different lib:
'assemblyIdentity name="System.Diagnostics.DiagnosticSource" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral"'

**Comment by smitpatel** [comment]
@codearoo - Can you run `dotnet --info` to find out which version of dotnet you are using? there was breaking change from 2.0 preview to RTM in dotnet around DiagnosticSource.

**Comment by codearoo** [comment]
@smitpatel 
`
.NET Command Line Tools (2.0.0)

Product Information:
 Version:            2.0.0
 Commit SHA-1 hash:  cdcd1928c9

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.15063
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.0.0\

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.0
  Build    : e8b8861ac7faf042c87a5c2f9f2d04c98b69f28d
`

**Comment by codearoo** [comment]
it's EntityFramework core project that seems to be throwing this.

**Comment by FransBouma** [comment]
I ran into the same issue. 

In vs2017 15.3.2, I have a classic csproj using .NET 4.6.2 with my unit tests for EF Core 2.0. When I create a DbContext derived type instance, I get:
```
System.IO.FileLoadException : Could not load file or assembly 'System.ValueTuple, Version=0.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
  Exception doesn't have a stacktrace
```

When I use the binding redirect I don't get the exception anymore but the test which caused this also doesn't run anymore, so the bindingredirect isn't really usable. 

Will try with a csproj in the new format to see whether that works...

**Comment by ErikEJ** [comment]
@FransBouma It works fine for me in a classic unit test csproj with the binding redirect:
https://github.com/ErikEJ/EntityFramework.SqlServerCompact/blob/master/test/EntityFramework.SqlServerCompact.FunctionalTests/app.config#L15

**Comment by FransBouma** [comment]
@ErikEJ hmm. You use nunit? As I couldn't get it to work whatever I tried. A new csproj format with these: https://github.com/dotnet/sdk/issues/1070 worked.

**Comment by ErikEJ** [comment]
@fransbouma as you can see from the tests i use xunit and execute with testdriven.net

**Comment by codearoo** [comment]
Just to update, it does work when I manually add this to app.config (have to create the file for projects that don't need one)
`<dependentAssembly> <assemblyIdentity name="System.ValueTuple" publicKeyToken="cc7b13ffcd2ddd51" /> <bindingRedirect oldVersion="0.0.0.0-4.0.2.0" newVersion="4.0.2.0" /> </dependentAssembly>`

But it's really kind of messed up.. For every project that uses the EntityFramework Core 2.0 stuff I have to add this secret sauce..  Microsoft isn't fixing this?

**Comment by evil-shrike** [comment]
Hi!

I'm sorry resurrecting this old closed issue but I'm having one of the mentioned error in my project: 

    Microsoft.NuGet.targets(178,5): error : Your project.json doesn't have a runtimes section. You should add '"runtimes": { "win": { } }' to your project.json and then re-run NuGet restore.

and really have no idea where to look after reading everything google found. All details are in the [SO question](https://stackoverflow.com/questions/46939387/). But briefly what was the reason of "Your project.json doesn't have a runtimes section" error? All subsequent suggestions on bindingRedirects puzzle me as I thought binding redirects happens at runtime but the error happens at build time.

**Comment by jessetabak** [comment]
I'm still having still problem, above solutions did not work, any suggestions?

EF Core version: 2.1.0-preview2-final
Database Provider: SQLite
Operating system: Win 10 1607
IDE: VS 2017 15.5.7

**Comment by smitpatel** [comment]
@jessetabak - Please file a new issue, with the description & repro steps of the issue you are seeing.
